### ISSUE TITLE
Open File Limit Error

### ISSUE THREAD

**Issue by RobertRosca** [European-XFEL/EXtra-data#2]
Open file limit per process may be hit for very large runs, leading to this error:

```
✘  roscar@max-exfl015 /gpfs/exfel/exp/MID/201931/p900091/raw/r0491  extra-data-validate .
Checking run directory: .
Traceback (most recent call last):
 File "/software/anaconda3/5.2/bin/extra-data-validate", line 10, in <module>
 File "/software/anaconda3/5.2/lib/python3.6/site-packages/extra_data/validation.py", line 222, in main
 File "/software/anaconda3/5.2/lib/python3.6/site-packages/extra_data/validation.py", line 163, in validate
 File "/software/anaconda3/5.2/lib/python3.6/site-packages/extra_data/validation.py", line 171, in run_checks
 File "/software/anaconda3/5.2/lib/python3.6/site-packages/extra_data/validation.py", line 191, in check_files_map
 File "/software/anaconda3/5.2/lib/python3.6/site-packages/extra_data/run_files_map.py", line 52, in __init__
 File "/software/anaconda3/5.2/lib/python3.6/site-packages/extra_data/run_files_map.py", line 86, in load
OSError: [Errno 24] Too many open files: './karabo_data_map.json
```

https://github.com/European-XFEL/EXtra-data/blob/e1f2b48e39f4627484bebc02045f554823f663ee/extra_data/run_files_map.py#L84-L90

Files may not be getting closed after they are looked at.

**Comment by takluyver** [comment]
Making a note for myself here: this is hopefully easy enough to fix for validate, but I want to check what happens with accessing the HDF5 files themselves - does HDF5 do anything to work around the issue, or do we need to be smarter with opening and closing files?

**Comment by tmichela** [comment]
That's strange. How the cache file was generated in the first place? (I tried to use `open_run` without using the cache and ran into the same issue...)

**Comment by tmichela** [comment]
got it: @ebadkamil just told me that he opened in the past this run per sequence (so not the full run). I guess then the cache file is used and did not run into this issue until running validate...

**Comment by takluyver** [comment]
HDF5 does not appear to do anything smart with this, so we'll have to work around it ourselves.

Do we have any examples of runs with that many files which are still on GPFS? The raw data from that MID proposal was migrated to dCache already, so testing with it can be extra slow.

**Comment by tmichela** [comment]
do we close this?

**Comment by egorsobolev** [comment]
The issue is still unsolved.
An example on GPFS: p002568, r0048, MID. This run contains 4055 files.
I read the BunchPattern or XGM data and read pulse indexes from every AGIPD panel. During the reading, all files open again and broke the limit.

There is an induced issue:
If the number of open files on the limit and I try to open new run replacing previous than I got the exception in `find_proposal` that proposal directory is not found:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/esobolev/proj/data-profile/extra_data/reader.py", line 1417, in open_run
  File "/home/esobolev/proj/data-profile/extra_data/read_machinery.py", line 207, in find_proposal
Exception: Couldn't find proposal dir for 'p002568'
```

It happens for me because I catch an exception on file limit, skip this run and go to the next. The solution for the user is to close all files (for example, call `del run`). But it is absolutely not obvious from the user side.

**Comment by takluyver** [comment]
Yeah, we mitigated this so you don't hit it as soon as you open a very large run, but you can still hit it if you want to access data from a lot of files.

As a further mitigation, you can raise the limit to the 'hard' limit of 4096, which might just be enough for that run:

```python
import resource
resource.setrlimit(resource.RLIMIT_NOFILE, (4096, 4096))
```

To fix this more properly, we'd probably need to keep track of which files have been accessed most recently, and automatically close older ones once it reaches a certain number. It's a bit messy, though, because you may have several runs open at once in the same process.

**Comment by takluyver** [comment]
This was properly addressed for 1.2 by PR #25.
